add spec infer related into prometheus metrics.

leiwen83 opened the pull request:
And add a new boost_ratio metric used to directly show how much spec infer would help in saving decoding steps.

leiwen83 commented:
cc @cadedaniel

cadedaniel commented:
will take a look Monday. btw, how is this different from system efficiency metric? (boost ratio == num_spec_tokens+1 * system efficiency?)

robertgshaw2-neuralmagic commented on vllm/engine/metrics.py:
```diff
@@ -59,7 +59,19 @@ def __init__(self, labelnames: List[str], max_model_len: int):
             documentation="CPU KV-cache usage. 1 means 100 percent usage.",
             labelnames=labelnames)
-
+        #   Speculative infer Status in %
```
Please add better descriptions.

Please name these:
- `vllm:spec_decode_system_efficiency`
- `vllm:spec_decode_boost_ratio`
- `vllm:spec_decode_draft_acceptance_rate`

What is the difference between these?

robertgshaw2-neuralmagic replied:
We prefer to use Counters >> Gauges

Is there a way these metrics could be expressed as Counters with the `rate` function used in PromQL to compute the rates?

leiwen83 replied:
maybe we could directly express the total emitted token, along with steps number? so that user could do the cal they want with those counters?

robertgshaw2-redhat reviewed (changes requested):
Please update the Grafana Dashboard

robertgshaw2-neuralmagic commented:
> will take a look Monday. btw, how is this different from system efficiency metric? (boost ratio == num_spec_tokens+1 * system efficiency?)

+1

robertgshaw2-neuralmagic commented:
Thanks for the contribution! It would be great to have these metrics flowing through prometheus!

leiwen83 commented:
> will take a look Monday. btw, how is this different from system efficiency metric? (boost ratio == num_spec_tokens+1 * system efficiency?)

the new boost_ratio would express more accurate expression at how much system is benefit from spec info, as there is case that spec info give no proposal, like no matching in ngram or seqlen+spec exceed over model length.

Furthermore, with the new dynamic spec coming https://github.com/vllm-project/vllm/issues/4565, the k would not be constant one, so that we may need accumulate actual token emitted comparing with the steps.

leiwen83 commented:
@cadedaniel @robertgshaw2-neuralmagic 
Any comment for the latest PR change? :)

cadedaniel commented:
asking @LiuXiaoxuanPKU if she has bandwidth to review the PR. the approach looks good to me, concerns are 1) we should make sure the top-level metrics make sense to users (not just to us as developers), 2) the naming of the metrics collection seems weird

robertgshaw2-neuralmagic commented:
I can review + merge tomorrow

leiwen83 commented:
Hi @robertgshaw2-neuralmagic @cadedaniel ,

How is going with the spec related metric, have we got the conclusion for how to make it happen? ;)
The metric is critical to us as a direct feedback reflecting how well current spec sys is doing.

cadedaniel commented:
thanks & sorry this slipped. I might have time tomorrow to finish review. cc @LiuXiaoxuanPKU and @comaniac who might have bandwidth.

cadedaniel commented on vllm/model_executor/layers/rejection_sampler.py:
```diff
@@ -318,10 +328,18 @@ def _create_output(
         self.num_emitted_tokens += (output_with_bonus_tokens != -1).sum()
         self.num_draft_tokens += batch_size * k
 
+        # batch_size here maybe 0, as there is case
+        # where no proposal is generated
+        self.num_specs += output_with_bonus_tokens.size()[0]
```
Is this right? even if there are no specs this can be nonzero

leiwen83 replied:
For the ngram case, if we don't match anything, then there would be no proposal.

cadedaniel replied:
OK. These metrics should work for draft model as well.

cadedaniel commented on vllm/model_executor/layers/rejection_sampler.py:
```diff
@@ -318,10 +328,18 @@ def _create_output(
             recovered_token_ids.mul(after_false_mask))
 
+        # concat spec and nonspec token
+        output_with_bonus_tokens = torch.cat(
+            [output_with_bonus_tokens, non_spec_token_ids])
```
`num_emitted_tokens` will include non-spec tokens, making the metric useless in capturing spec system efficiency (100% efficiency is defined every token accepted plus the bonus token)

cadedaniel commented on vllm/spec_decode/spec_decode_worker.py:
```diff
@@ -311,20 +311,20 @@ def _verify_tokens(
-        non_spec_token_ids = non_spec_token_ids.expand(-1, max_proposal_len +
-                                                       1).clone()
-        non_spec_token_ids[:, 1:] = -1
-        accepted_token_ids = torch.cat(
-            [accepted_token_ids, non_spec_token_ids])
```
btw this cat should stay here; the format required for the cat is determined here `original_indices = spec_indices + non_spec_indices`

cadedaniel commented on vllm/spec_decode/metrics.py:
```diff
@@ -162,6 +168,7 @@ def _collect_rejsample_metrics(
 
         return SpecDecodeWorkerMetrics(
             num_spec_tokens=k,
+            num_specs=self._aggregate_num_specs,
```
we can calculate this with `draft_tokens // k`, don't need to record

cadedaniel replied:
see `get_max_num_emitted_tokens`

cadedaniel reviewed (commented):
I have bandwidth to review this now. The Prometheus stuff looks good to me, have concerns on the metric definition and how we collect them.

wenlei03 committed:
add spec infer related into prometheus metrics.

And add a new boost_ratio metric used to directly show how much spec
infer would help in saving decoding steps.

Signed-off-by: Lei Wen <wenlei03@qiyi.com>

leiwen83 commented:
@cadedaniel 
I submit a rebased PR, which keep the concat logic as before. num_spec is made to aggregate "batch_size*k" number.

wenlei03 committed:
update
